include/fir/Type.h

```diff
@@ -0,0 +1,274 @@
+// limitations under the License.
+
+#ifndef FIR_TYPE_H
+#define FIR_TYPE_H
+
+#include "mlir/IR/Types.h"
+
+namespace llvm {
+class StringRef;
+template<typename> class ArrayRef;
+class hash_code;
+}
+
+namespace fir {
+
+class FIROpsDialect;
+
+using KindTy = int;
+
+namespace detail {
+struct BoxTypeStorage;
+struct BoxCharTypeStorage;
+struct BoxProcTypeStorage;
+struct CharacterTypeStorage;
+struct CplxTypeStorage;
+struct DimsTypeStorage;
+struct FieldTypeStorage;
+struct HeapTypeStorage;
+struct IntTypeStorage;
+struct LogicalTypeStorage;
+struct PointerTypeStorage;
+struct RealTypeStorage;
+struct RecordTypeStorage;
+struct ReferenceTypeStorage;
+struct SequenceTypeStorage;
+struct TypeDescTypeStorage;
+}
+
+enum TypeKind {
+  // The enum starts at the range reserved for this dialect.
+  FIR_TYPE = mlir::Type::FIRST_FIR_TYPE,
+  FIR_BOX,
+  FIR_BOXCHAR,
+  FIR_BOXPROC,
+  FIR_CHARACTER,  // intrinsic
+  FIR_COMPLEX,  // intrinsic
+  FIR_DERIVED,  // derived
+  FIR_DIMS,
+  FIR_FIELD,
+  FIR_HEAP,
+  FIR_INT,  // intrinsic
+  FIR_LOGICAL,  // intrinsic
+  FIR_POINTER,
+  FIR_REAL,  // intrinsic
+  FIR_REFERENCE,
+  FIR_SEQUENCE,
+  FIR_TYPEDESC,
+};
+
+bool isa_fir_type(mlir::Type);
+bool isa_std_type(mlir::Type t);
+bool isa_fir_or_std_type(mlir::Type t);
+
+template<typename A, unsigned Id> struct IntrinsicTypeMixin {
+  constexpr static bool kindof(unsigned kind) { return kind == getId(); }
+  constexpr static unsigned getId() { return Id; }
+};
+
+class CharacterType
+  : public mlir::Type::TypeBase<CharacterType, mlir::Type,
+        detail::CharacterTypeStorage>,
+    public IntrinsicTypeMixin<CharacterType, TypeKind::FIR_CHARACTER> {
+public:
+  using Base::Base;
+  static CharacterType get(mlir::MLIRContext *ctxt, KindTy kind);
+  int getSizeInBits() const;
+  KindTy getFKind() const { return getSizeInBits() / 8; }
+};
+
+class IntType
+  : public mlir::Type::TypeBase<IntType, mlir::Type, detail::IntTypeStorage>,
+    public IntrinsicTypeMixin<IntType, TypeKind::FIR_INT> {
+public:
+  using Base::Base;
+  static IntType get(mlir::MLIRContext *ctxt, KindTy kind);
+  int getSizeInBits() const;
+  KindTy getFKind() const { return getSizeInBits() / 8; }
+};
+
+class LogicalType
+  : public mlir::Type::TypeBase<LogicalType, mlir::Type,
+        detail::LogicalTypeStorage>,
+    public IntrinsicTypeMixin<LogicalType, TypeKind::FIR_LOGICAL> {
+public:
+  using Base::Base;
+  static LogicalType get(mlir::MLIRContext *ctxt, KindTy kind);
+  int getSizeInBits() const;
+  KindTy getFKind() const { return getSizeInBits() / 8; }
+};
+
+class RealType
+  : public mlir::Type::TypeBase<RealType, mlir::Type, detail::RealTypeStorage>,
+    public IntrinsicTypeMixin<RealType, TypeKind::FIR_REAL> {
+public:
+  using Base::Base;
+  static RealType get(mlir::MLIRContext *ctxt, KindTy kind);
+  int getSizeInBits() const;
+  KindTy getFKind() const { return getSizeInBits() / 8; }
+};
+
+class CplxType
```

Comment:
Any reason not to spell out Complex?

Reply:
Yes. It conflicts with mlir::ComplexType.

Reply:
Why not let the namespace help you use a nicer name?   fir:: is not mlir::

Reply:
It's not possible here. I've tried it.

Reply:
Bummer.